File: backend/app/services/youtube_service.py
from youtubesearchpython import VideosSearch


async def search_videos(query: str, max_results: int = 1):
    try:
        videos_search = VideosSearch(query, limit=max_results)
        results = videos_search.result()
        
        # Normalize to match expected format
        videos = []
        for item in results['result']:
            videos.append({
                "id": {"videoId": item['id']},
                "snippet": {
                    "title": item['title'],
                    "thumbnails": {"high": {"url": item['thumbnails'][0]['url']}},
                    "channelTitle": item['channel']['name']
                }
            })
        return videos
    except Exception as e:
        print(f"YouTube Search Error: {e}")
        return []


async def get_video_metadata(video_id: str):
    try:
        videos_search = VideosSearch(video_id, limit=1)
        results = videos_search.result()
        if results['result']:
            item = results['result'][0]
            return {
                "id": item['id'],
                "snippet": {
                    "title": item['title'],
                    "description": item.get('descriptionSnippet', [{}])[0].get('text', ''),
                    "thumbnails": {"high": {"url": item['thumbnails'][0]['url']}},
                },
                "contentDetails": {
                    "duration": item['duration']
                }
            }
        return None
    except Exception as e:
        print(f"YouTube Metadata Error: {e}")
        return None

File: backend/src/services/youtubeService.ts
import { google } from 'googleapis';

const youtube = google.youtube('v3');
const API_KEY = process.env.YOUTUBE_API_KEY;

// Educational keywords appended to subtopic for YouTube search (per spec)
const EDUCATIONAL_KEYWORDS = 'tutorial education lesson learn';

// Simple in-memory cache for ETag-like behavior
const videoCache = new Map<string, { etag: string, data: any, timestamp: number }>();
const CACHE_TTL = 1000 * 60 * 60 * 24; // 24 hours

/**
 * Automated Video Retrieval (per spec):
 * - Subtopic + educational keywords sent as query to YouTube Data API v3.
 * - search.list with order=relevance, maxResults=30.
 * - Second request with videos.list part=statistics to get likeCount.
 * - Top 5 videos shortlisted by likeCount. Repeated per subtopic by caller.
 */
export const searchVideos = async (subtopic: string): Promise<any[]> => {
    if (!API_KEY) {
        throw new Error("YOUTUBE_API_KEY is not set");
    }

    try {
        const query = [subtopic.trim(), EDUCATIONAL_KEYWORDS].filter(Boolean).join(' ');

        // 1) Search: order=relevance, maxResults=30
        const searchResponse = await youtube.search.list({
            key: API_KEY,
            part: ['snippet'],
            q: query,
            type: ['video'],
            order: 'relevance',
            maxResults: 30,
            videoDuration: 'medium',
            relevanceLanguage: 'en'
        });

        const searchItems = searchResponse.data.items || [];
        if (searchItems.length === 0) return [];

        const videoIds = searchItems
            .map((item: any) => item.id?.videoId)
            .filter(Boolean);
        if (videoIds.length === 0) return [];

        // 2) Second request: part=statistics (and snippet for consistent shape)
        const statsResponse = await youtube.videos.list({
            key: API_KEY,
            part: ['snippet', 'statistics'],
            id: videoIds
        });

        const videosWithStats = (statsResponse.data.items || []) as any[];
        const withLikeCount = videosWithStats.map((v) => ({
            ...v,
            likeCount: parseInt(String(v.statistics?.likeCount || 0), 10) || 0
        }));
        withLikeCount.sort((a, b) => b.likeCount - a.likeCount);

        const top5 = withLikeCount.slice(0, 5);
        return top5.map((v) => ({
            id: { videoId: String(v.id) },
            snippet: v.snippet || { title: '', thumbnails: { high: { url: '' } } }
        }));
    } catch (error) {
        console.error("YouTube Search Error:", error);
        return [];
    }
};

export const getVideoMetadata = async (videoId: string) => {
    if (!API_KEY) throw new Error("YOUTUBE_API_KEY is not set");

    const cached = videoCache.get(videoId);
    if (cached && (Date.now() - cached.timestamp < CACHE_TTL)) {
        return cached.data;
    }

    // ETag logic integration (simulated since we aren't storing the actual YouTube ETag from their header yet, 
    // but we check our cache first).
    // Real ETag usage with YouTube API involves sending the `ifNoneMatch` header.
    // Let's implement that if we had the real etag.

    try {
        const response = await youtube.videos.list({
            key: API_KEY,
            part: ['snippet', 'contentDetails', 'statistics'],
            id: [videoId]
        });

        const data = response.data.items?.[0];
        if (data) {
            videoCache.set(videoId, {
                etag: response.data.etag || 'no-etag',
                data: data,
                timestamp: Date.now()
            });
        }
        return data;
    } catch (error) {
        console.error("YouTube Metadata Error:", error);
        return null;
    }
};

// Fallback for when we want to just search generic "Uploads" from a channel if we knew it
// Implementing the requested "PlaylistItems" preference logic:
export const searchChannelVideos = async (channelId: string, query: string) => {
    // This uses 1 quota unit per page vs 100 for search
    // But requires knowing the channel ID. 
    // We will stick to refined search.list for now as we search globally.
    return [];
};
